File: app/rcc/scripts/mqtt/lua/influxdb.lua
--
-- InfluxDB line-protocol publisher.
--
-- Line protocol:
--   measurement[,tag=value...] field=value[,field=value...] [timestamp]
--
-- Pipe this topic into a Telegraf mqtt_consumer input with the
-- influx_v2 data_format and it lands directly in your bucket.
--

local MEASUREMENT = "serial_studio"
local TAGS        = "device=dev01"
local FIELDS      = { "temperature", "humidity", "pressure" }

local function split(s, sep)
  local out, i = {}, 1
  for part in string.gmatch(s, "([^" .. sep .. "]+)") do
    out[i] = part
    i = i + 1
  end
  return out
end

function mqtt(frame)
  local parts  = split(tostring(frame), ",")
  local fields = {}

  for i, key in ipairs(FIELDS) do
    local v = tonumber(parts[i])
    if v then
      table.insert(fields, key .. "=" .. v)
    end
  end

  if #fields == 0 then
    return nil
  end

  -- Nanosecond timestamp; os.time() is seconds so multiply by 1e9
  local ts = string.format("%d", os.time() * 1000000000)
  return MEASUREMENT .. "," .. TAGS .. " "
       .. table.concat(fields, ",") .. " " .. ts .. "\n"
end

File: app/rcc/scripts/mqtt/js/influxdb.js
/*
 * InfluxDB line-protocol publisher.
 *
 * Line protocol:
 *   measurement[,tagKey=tagValue...] fieldKey=fieldValue[,...] [timestamp]
 *
 * Pipe this MQTT topic into a Telegraf MQTT consumer with the
 * `mqtt_consumer` input + `influx_v2` data_format and it lands directly
 * in your bucket.
 */

var MEASUREMENT = "serial_studio";
var TAGS = "device=dev01";
var FIELDS = ["temperature", "humidity", "pressure"];

function mqtt(frame) {
  var parts = String(frame).split(",");
  var fields = [];

  for (var i = 0; i < FIELDS.length && i < parts.length; ++i) {
    var v = parseFloat(parts[i]);
    if (!isFinite(v))
      continue;

    fields.push(FIELDS[i] + "=" + v);
  }

  if (fields.length === 0)
    return null;

  // Nanosecond timestamp; Date.now() is ms so multiply by 1e6
  var ts = (Date.now() * 1e6).toFixed(0);
  return MEASUREMENT + "," + TAGS + " " + fields.join(",") + " " + ts + "\n";
}
